packetizer.go

```diff
@@ -1,64 +1,89 @@
 package rpc
 
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"sync"
+
+	"github.com/ugorji/go/codec"
+)
+
 type packetizer interface {
-	Packetize() error
+	NextFrame() (RPCMessage, error)
 }
 
 type packetHandler struct {
-	receiver receiver
-	dec      byteReadingDecoder
+	dec           decoder
+	reader        io.Reader
+	frameDecoders decoderPool
+	protocols     *protocolHandler
+	calls         *callContainer
+	mtx           sync.Mutex
 }
 
-func newPacketHandler(r receiver, dec byteReadingDecoder) *packetHandler {
+func newPacketHandler(reader io.Reader, protocols *protocolHandler, calls *callContainer) *packetHandler {
 	return &packetHandler{
-		receiver: r,
-		dec:      dec,
+		reader:        reader,
+		dec:           codec.NewDecoder(reader, newCodecMsgpackHandle()),
+		frameDecoders: makeDecoderPool(),
+		protocols:     protocols,
+		calls:         calls,
 	}
 }
 
-func (p *packetHandler) getFrame() (int, error) {
-	var l int
-
-	err := p.dec.Decode(&l)
-
-	return l, err
-}
-
-func (p *packetHandler) getMessage(l int) (err error) {
-	// TODO currently tossing out `l` above. We should either validate it or
-	// not pass it in at all.
-	var b byte
-
-	if b, err = p.dec.ReadByte(); err != nil {
-		return err
+func (p *packetHandler) NextFrame() (RPCMessage, error) {
+	bytes, err := p.loadNextFrame()
+	if err != nil {
+		return nil, err
+	}
+	if len(bytes) < 1 {
+		return nil, fmt.Errorf("invalid frame size: %d", len(bytes))
 	}
-	nb := int(b)
+
+	// Attempt to read the fixarray
+	nb := int(bytes[0])
 
 	// Interpret the byte as the length field of a fixarray of up
 	// to 15 elements: see
 	// https://github.com/msgpack/msgpack/blob/master/spec.md#formats-array
 	// . Do this so we can decode directly into the expected
 	// fields without copying.
-	if nb >= 0x91 && nb <= 0x9f {
-		err = p.receiver.Receive(nb - 0x90)
-	} else {
-		err = NewPacketizerError("wrong message structure prefix (%d)", nb)
+	if nb < 0x91 || nb > 0x9f {
+		return nil, NewPacketizerError("wrong message structure prefix (%d)", nb)
 	}
+	dec := p.frameDecoders.getDecoder(bytes[1:])
```

Comment:
I don't quite understand why you need the decoder pool -- this is also called synchronously, so why can't you just use a single decoder?